modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/model/listener/HeadlessBuilderObjectEntryModelListener.java

```diff
@@ -0,0 +1,154 @@
+ * details.
+ */
+
+package com.liferay.headless.builder.internal.model.listener;
+
+import com.liferay.headless.builder.application.APIApplication;
+import com.liferay.headless.builder.application.provider.APIApplicationProvider;
+import com.liferay.headless.builder.application.publisher.APIApplicationPublisher;
+import com.liferay.object.model.ObjectDefinition;
+import com.liferay.object.model.ObjectEntry;
+import com.liferay.object.service.ObjectDefinitionLocalService;
+import com.liferay.object.service.ObjectEntryLocalService;
+import com.liferay.portal.kernel.exception.ModelListenerException;
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.transaction.TransactionCommitCallbackUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.io.Serializable;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+@Component(service = ModelListener.class)
+public class HeadlessBuilderObjectEntryModelListener
+	extends BaseModelListener<ObjectEntry> {
+
+	@Override
+	public void onAfterCreate(ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	@Override
+	public void onAfterUpdate(
+			ObjectEntry originalObjectEntry, ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	private long _getAPIApplicationId(ObjectEntry objectEntry) {
+		if (_isObjectDefinition("L_API_APPLICATION", objectEntry)) {
+			return objectEntry.getObjectEntryId();
+		}
+		else if (_isObjectDefinition("L_API_SCHEMA", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPISchemas_c_apiApplicationId");
+		}
+		else if (_isObjectDefinition("L_API_ENDPOINT", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPIEndpoints_c_apiApplicationId");
+		}
+
+		throw new UnsupportedOperationException("Invalid object definition");
+	}
+
+	private boolean _isObjectDefinition(
+		String objectDefinitionExternalReferenceCode, ObjectEntry objectEntry) {
+
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectEntry.getObjectDefinitionId());
+
+		return Objects.equals(
+			objectDefinition.getExternalReferenceCode(),
+			objectDefinitionExternalReferenceCode);
+	}
+
+	private void _schedulePublication(ObjectEntry objectEntry) {
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectEntry.getObjectDefinitionId());
+
+		if (StringUtil.startsWith(
```

Comment:
I think this if clause could be dangerous, as a different modifiable system object could called starting with `L_API_` and that would cause an `UnsupportedOperationException`.

Maybe we could manage in a different way as we have different if /else clauses in the `_getAPIApplicationId` method

Reply:
Yes, but we (Liferay) control those prefixes and we should not make that mistake. Do you have an idea on how to do this better?

Reply:
Good point, we didn't see it. I think that we can manage the `APIApplicationId` with the if/else statements

Reply:
The simplest thing would be to have a list of the whole names.

Reply:
What about something like:

```
	private Long _getAPIApplicationId(ObjectEntry objectEntry) {
		ObjectDefinition objectDefinition =
			_objectDefinitionLocalService.fetchObjectDefinition(
				objectEntry.getObjectDefinitionId());

		String externalReferenceCode = objectDefinition.getExternalReferenceCode();

		if (StringUtil.equals(externalReferenceCode, "L_API_APPLICATION")) {
			return objectEntry.getObjectEntryId();
		}
		else if (StringUtil.equals(externalReferenceCode, "L_API_SCHEMA")) {
			Map<String, Serializable> values = objectEntry.getValues();

			return (long)values.get(
				"r_apiApplicationToAPISchemas_c_apiApplicationId");
		}
		else if (StringUtil.equals(externalReferenceCode, "L_API_ENDPOINT")) {
			Map<String, Serializable> values = objectEntry.getValues();

			return (long)values.get(
				"r_apiApplicationToAPIEndpoints_c_apiApplicationId");
		}

		if (_log.isDebugEnabled()) {
			_log.debug("......");
		}

		return null;
	}
```

```
	private void _schedulePublication(ObjectEntry objectEntry) {

		Long apiApplicationId = _getAPIApplicationId(objectEntry);

		if (apiApplicationId == null) {
			return;
		}

		_pendingAPIApplications.add(apiApplicationId);
```

Reply:
Thanks @ccorreagg!! I will take a view at refactor it!!

Reply:
Do you like that @4lejandrito ?

Reply:
Yes, he likes it :smile:

Reply:
https://github.com/liferay-headless/liferay-portal/pull/1321/commits/78cfb2a2138d38eb41e317cdf53caaeb2f444c5c